Add beekeeper

Minecraftschurli opened the pull request:
# Changes proposed in this pull request:
- add Beekeeper colonist

- [x] mechanics (harvesting/breeding)
- [x] hut recipe
- [x] hut block model & texture
- [x] scepter texture
- [ ] translations
- [ ] schematics (beekeeper hut has no special requirement)

Please review

Minecraftschurli committed:
Initial beekeeper commit

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=5127) <br/>All committers have signed the CLA.

Raycoms commented:
Check out my blog: https://hive.blog/utopian-io/@raycoms/minecolonies--the-crusher

What the necessary steps are for a worker (where things have to be registered too etc)

Minecraftschurli commented:
thanks (I just commited the "empty" classes)

Minecraftschurli committed:
add more beekeeper classes

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
start on ai

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Minecraftschurli committed:
implement beekeeper ai

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
implement beekeeper ai

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add window

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Minecraftschurli committed:
finish (missing:
- translations
- scepter texture
- schematics
- hut recipe)

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli commented:
still missing:
- translations
- ~~scepter texture~~
- schematics (beekeeper hut has no special requirement)
- ~~hut recipe~~

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

ravenbuilder934 commented:
> still missing:
> 
> * translations
> * scepter texture
> * schematics (beekeeper has no special requirement)
> * hut recipe

Wouldn't the beekeeper need beehives as part of the schematic? What do you mean by beekeeper has no special requirement?

Minecraftschurli commented:
the schematic has no special requirement and beehives/nests can be added via the scepter (like lumberjack)

Minecraftschurli committed:
run datagen

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add recipe

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add scepter texture

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Raycoms commented on src/api/java/com/minecolonies/api/entity/ai/statemachine/states/AIWorkerState.java:
```diff
@@ -533,7 +533,12 @@
+    /*
+### Plantation related ###
+     */
+    BEEKEEPER_HARVEST(true);
```
Wrong jdoc

Minecraftschurli replied:
oops

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutBeekeeper.java:
```diff
@@ -0,0 +1,27 @@
+import com.minecolonies.api.colony.buildings.ModBuildings;
+import com.minecolonies.api.colony.buildings.registry.BuildingEntry;
+
+public class BlockHutBeekeeper extends AbstractBlockHut<BlockHutBeekeeper> {
```
formatting is off

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java:
```diff
@@ -0,0 +1,258 @@
+     */
+    private static final String BEEKEEPER = "beekeeper";
+
+    private static final String TAG_HIVES              = "hives";
```
NBTtag constant class please

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java:
```diff
@@ -0,0 +1,258 @@
+     * List of hives.
+     */
+    private Set<BlockPos> hives             = new HashSet<>();
+    private boolean       harvestHoneycombs = true;
```
jdoc missing

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java:
```diff
@@ -0,0 +1,258 @@
+        return nbt;
+    }
+
+    @Override
```
I don't think the worker needs any recipes

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java:
```diff
@@ -0,0 +1,258 @@
+        buf.writeBoolean(harvestHoneycombs);
+    }
+
+    public Set<BlockPos> getHives()
```
@Override for overriden methods, else javadoc

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/jobs/JobBeekeeper.java:
```diff
@@ -0,0 +1,64 @@
+        return new EntityAIWorkBeekeeper(this);
+    }
+
+    @Override
```
not needed if empty

Raycoms replied:
You should remove the entire deserializeMethod

Raycoms commented on src/main/java/com/minecolonies/coremod/network/messages/server/colony/building/beekeeper/BeekeeperScepterMessage.java:
```diff
@@ -0,0 +1,83 @@
+import static com.minecolonies.api.util.constant.NbtTagConstants.TAG_POS;
+
+/**
+ * Message to set the lumberjack scepter in the player inventory.
```
jdoc

Raycoms commented on src/main/java/com/minecolonies/coremod/network/messages/server/colony/building/beekeeper/BeekeeperSetHarvestHoneycombsMessage.java:
```diff
@@ -0,0 +1,62 @@
+
+public class BeekeeperSetHarvestHoneycombsMessage extends AbstractBuildingServerMessage<BuildingBeekeeper>
+{
+    private boolean harvestHoneycombs;
```
jdoc

Raycoms commented on src/main/java/com/minecolonies/coremod/items/ItemScepterBeekeeper.java:
```diff
@@ -0,0 +1,116 @@
+        // if server world, do nothing
+        if (useContext.getWorld().isRemote)
+        {
+            return ActionResultType.FAIL;
```
Should probably be pass and not fail.

Minecraftschurli replied:
it's copied from lumberjack so I thought this is right

Raycoms replied:
Ah, then you can keep it. There it works fine.

Raycoms commented on src/main/java/com/minecolonies/coremod/items/ItemScepterBeekeeper.java:
```diff
@@ -0,0 +1,116 @@
+            save(useContext.getPlayer(), building, compound);
+        }
+
+        return super.onItemUse(useContext);
```
Might be easier to save them directly in the building each time already.

Raycoms replied:
That would also save you from writing positions in the item

Minecraftschurli replied:
yea that's way better (was thinking too complicated when I did it)

Minecraftschurli committed:
add changes requested by Raycoms

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
fix merge conflict for class ModBlocks

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
remove test schematics

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
fix blockHutArchery being used twice in getHuts

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutBeekeeper.java:
```diff
@@ -0,0 +1,28 @@
+     * @return Name of the block.
+     */
+    @Override
+    public String getName() {
```
formatting slightly off

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutBeekeeper.java:
```diff
@@ -0,0 +1,28 @@
+import com.minecolonies.api.colony.buildings.ModBuildings;
+import com.minecolonies.api.colony.buildings.registry.BuildingEntry;
+
+public class BlockHutBeekeeper extends AbstractBlockHut<BlockHutBeekeeper>
```
jdoc

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java:
```diff
@@ -0,0 +1,259 @@
+    @Override
+    public Skill getPrimarySkill()
+    {
+        return Skill.Stamina; //TODO
```
Dexterity -> primary
Adaptability -> Secondary

Were planned for this worker

Minecraftschurli replied:
ok (that's why I put the todo there)

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/jobs/JobBeekeeper.java:
```diff
@@ -0,0 +1,60 @@
+import com.minecolonies.coremod.entity.ai.citizen.beekeeper.EntityAIWorkBeekeeper;
+import net.minecraft.nbt.INBT;
+
+public class JobBeekeeper extends AbstractJob
```
jdoc

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1495,6 +1499,40 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+            return true;
+        }
+
+        final List<ItemStack> stacks = tag.getAllElements().stream().map(itemIn -> new ItemStack(itemIn, count)).collect(Collectors.toList());
```
That is quite the creative way to make a tag based request. Although I'm unsure we shouldn't create just a TagBased request type new instead of doing it the hacky way.

Raycoms commented on src/main/java/com/minecolonies/coremod/items/ItemScepterBeekeeper.java:
```diff
@@ -0,0 +1,88 @@
+        if (useContext.getWorld().getBlockState(useContext.getPos()).getBlock() instanceof BeehiveBlock)
+        {
+
+            final Collection<BlockPos> positions = building.getHives();
```
I don't like the code style of this. We should have a setter for this on the building instead of getting a mutable list and add them by reference.

Minecraftschurli replied:
you mean add and remove proxy methods?

Raycoms replied:
yes

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java:
```diff
@@ -0,0 +1,460 @@
+        );
+    }
+
+    public boolean maxAnimals(final List<BeeEntity> allAnimals)
```
private, jdoc, and better name

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java:
```diff
@@ -0,0 +1,460 @@
+            }
+        }
+
+        checkIfRequestForItemExistOrCreateAsynch(new ItemStack(Items.GLASS_BOTTLE));
```
Possibly, when there are a bunch of glass bottles and flowers in the inventory they shouldn't request

Minecraftschurli replied:
I think checkIfRequestForItemExistOrCreateAsynch does already check the inventory

Minecraftschurli replied:
from checkIfRequestForItemExistOrCreateAsynch jdoc
> Check if a stack has been requested already or is in the inventory. If not in the inventory and not requested already, create request

Raycoms replied:
ah okay

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java:
```diff
@@ -0,0 +1,460 @@
+
+        final List<BeeEntity> bees = new ArrayList<>(searchForAnimals());
+
+        if (bees.isEmpty())
```
Should probably set an interaction here, if they can't find bees within a time they gotta tell the player about this.

Minecraftschurli replied:
how should I implement this? with a counter variable?

Raycoms replied:
Better probably if hive is empty. Then add the interaction.

Minecraftschurli committed:
continue adding requested changes

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
fix errors that resulted from upstream changes

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add proxy methods for adding/removing hives/nests to/from the beekeeper

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
Merge branch 'version/1.15' into feature/beekeeper

Raycoms committed:
Div text (#5175)

Some minor fixes
Add full diversity textures to minecolonies entities

PowerSchill committed:
Feature/recipe tags (#5058)

Introduces a system for adding and excluding recipes based on datapack tags

lunadragnstar committed:
Schematics115 (#5189)

Schematic Deco Fixes

PowerSchill committed:
Removes fulfill and cancel from clipboard detail view (#5180)

Removes fulfill from clipboard detail view

Raycoms committed:
Improve spawner in raids, improve ticks (#5199)

Tracks spawner position in raid event and regularly checks them
Also now uses the O(1) methods of racks better for the minimum stock ticks.

someaddons reviewed (dismissed):
Looks ok

ToMe25 committed:
Add Stonebrick slabs and stairs to the list of path blocks. (#5219)

Add Stone Brick slabs and stairs to the list of path blocks.
This makes them prefer walking on them, and get a speed boost.

Minecraftschurli committed:
add a Tag based request (#5185)

this PR adds a Tag based request as proposed by @Raycoms in #5127

Minecraftschurli committed:
transition to tag requests

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
implement skill based features

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add EntityAIWorkBeekeeper jdoc

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Raycoms commented:
@Minecraftschurli can you solve the conflicts?

Minecraftschurli committed:
Merge remote-tracking branch 'upstream/version/1.15' into feature/beekeeper

# Conflicts:
#	build.properties
#	src/api/java/com/minecolonies/api/blocks/ModBlocks.java
#	src/api/java/com/minecolonies/api/colony/buildings/ModBuildings.java
#	src/api/java/com/minecolonies/api/colony/jobs/ModJobs.java
#	src/api/java/com/minecolonies/api/util/constant/NbtTagConstants.java
#	src/main/generated/resources/assets/minecolonies/sounds.json
#	src/main/java/com/minecolonies/apiimp/initializer/ModBlocksInitializer.java
#	src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java
#	src/main/java/com/minecolonies/apiimp/initializer/ModJobsInitializer.java
#	src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingCrafter.java
#	src/main/resources/META-INF/mods.toml
#	src/main/resources/assets/minecolonies/schematics/wooden/archery2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/archery3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/archery4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/archery5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/baker1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/baker2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/baker3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/baker4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/baker5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barracks1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barracks2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barracks3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barracks4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barracks5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/barrackstower3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/blacksmith1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/blacksmith2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/blacksmith3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/blacksmith4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/blacksmith5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/builder1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/builder2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/builder3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/builder4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/builder5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/chickenherder1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/chickenherder2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/chickenherder3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/chickenherder4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/chickenherder5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/citizen1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/citizen2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/citizen3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/citizen4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/citizen5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/composter1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/composter2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/composter3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/composter4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cook1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cook2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cook3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cook4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cook5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cowboy1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cowboy2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cowboy3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cowboy4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/cowboy5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/crusher1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/crusher2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/crusher3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/crusher4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/deliveryman1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/deliveryman2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/deliveryman3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/deliveryman4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/deliveryman5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/dyer1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/dyer2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/dyer3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/enchanter1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/enchanter2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/enchanter3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/enchanter4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/enchanter5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/farmer1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/farmer2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/farmer3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/farmer4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/farmer5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fisherman1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fisherman2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fisherman3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fisherman4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fisherman5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fletcher1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fletcher2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/fletcher3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/florist1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/florist2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/florist3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/florist4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/florist5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/glassblower1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/glassblower4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/guardtower1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/guardtower2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/guardtower3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/guardtower4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/guardtower5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/hospital1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/hospital2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/library1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/library2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/library4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/library5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/lumberjack1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/lumberjack2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/lumberjack3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/lumberjack4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/mechanic1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/mechanic2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/mechanic5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/miner1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/miner2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/miner3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/miner4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/miner5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/plantation1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/plantation2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/plantation3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/plantation4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/plantation5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sawmill1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sawmill2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sawmill3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sawmill4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sawmill5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/school2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/school3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/shepherd1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/shepherd2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/shepherd3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sifter1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sifter2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sifter3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/sifter4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/smeltery2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/smeltery3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/smeltery4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/smeltery5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonemason1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonemason5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonesmeltery1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonesmeltery2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonesmeltery3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/stonesmeltery4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/swineherder1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/swineherder2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/tavern1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/tavern2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/tavern3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/university1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/university2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/university3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/university4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/university5.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/warehouse1.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/warehouse2.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/warehouse3.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/warehouse4.blueprint
#	src/main/resources/assets/minecolonies/schematics/wooden/warehouse5.blueprint
#	src/main/resources/assets/minecolonies/textures/entity/citizen/default/citizenmale1_a.png
#	src/main/resources/assets/minecolonies/textures/entity/citizen/default/citizenmale1_b.png
#	src/main/resources/assets/minecolonies/textures/entity/citizen/default/citizenmale1_d.png

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1505,6 +1507,39 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+     * @param tag the requested tag.
+     * @return true if in the inventory, else false.
+     */
+    public boolean checkIfRequestForTagExistOrCreateAsynch(@NotNull final Tag<Item> tag, final int count)
```
you should be able to simplify this one here now.

Minecraftschurli replied:
it already is

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java:
```diff
@@ -0,0 +1,496 @@
+            final int numOfAnimals = allBees.size();
+            final int maxAnimals = getOwnBuilding().getBuildingLevel() * BEES_PER_LEVEL;
+
+            return (numOfAnimals + numOfBeesInHive) > maxAnimals;
```
>= ?

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1505,6 +1507,39 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+            return true;
+        }
+
+        if (getOwnBuilding().getOpenRequestsOfTypeFiltered(worker.getCitizenData(), TypeConstants.DELIVERABLE,
```
You can here filter it by TagRequest instead of Deliverable already, that's what I meant

Minecraftschurli replied:
So I should create a new entry in TypeConstants

Raycoms replied:
yes

Minecraftschurli commented:
So I should create a new entry in TypeConstants

Minecraftschurli committed:
implement changes requested by @Raycoms

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add interaction if the beekeeper has no bees as suggested by @Raycoms

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/5127/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

someaddons commented on src/main/java/com/minecolonies/apiimp/initializer/InteractionValidatorInitializer.java:
```diff
@@ -177,6 +180,18 @@ public static void init()
 
+        InteractionValidatorRegistry.registerStandardPredicate(new TranslationTextComponent(NO_BEES),
+          citizen -> citizen.getWorkBuilding() instanceof BuildingBeekeeper &&
+                       EntityAIWorkBeekeeper.searchForAnimals(citizen.getColony().getWorld(), (BuildingBeekeeper) citizen.getWorkBuilding()).isEmpty() &&
```
simplify this to a getHives().isEmpty() on the building, that way you don't use expensive area checks or accidentally load chunks. Let the AI do the checking

Minecraftschurli replied:
well but then the interaction doesn't check the bees outside the hive

someaddons replied:
isnt that fine? like do there have to be bees outside

Minecraftschurli replied:
well this is to tell the player that the beekeeper has no bees but if there are bees around the hive they are there so players would see it as a bug

someaddons replied:
doesnt this also happen when the bee's happen to be inside the hives?

Minecraftschurli replied:
no that's what the other part is for
```java
((BuildingBeekeeper) citizen.getWorkBuilding())
                         .getHives()
                         .stream()
                         .map(citizen.getColony().getWorld()::getTileEntity)
                         .filter(tileEntity -> tileEntity instanceof BeehiveTileEntity)
                         .map(tileEntity -> (BeehiveTileEntity) tileEntity)
                         .mapToInt(BeehiveTileEntity::getBeeCount)
                         .sum() <= 0)
```

Raycoms replied:
only the hives is fine. About the bees the players can't do anything what matters is if there are no hives marked with the tool

Minecraftschurli replied:
ok so change it from being no bees to no hives

someaddons reviewed (commented):
One last change needed then good to go

Minecraftschurli committed:
split interaction for the beekeeper into has no bees and has no hives

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Minecraftschurli committed:
add jdoc

Signed-off-by: Minecraftschurli <minecraftschurli@gmail.com>

Asherslab reviewed (dismissed):
looks good to me

Minecraftschurli committed:
Merge remote-tracking branch 'upstream/version/1.15' into feature/beekeeper

# Conflicts:
#	src/api/java/com/minecolonies/api/util/constant/TranslationConstants.java
